### PR TITLE
LiftSim A2C baseline

### PR DESCRIPTION
**Pull request by zenghsh3**
- add a2c baseline of liftsim env

### CODE REVIEW

File: examples/LiftSim_baseline/A2C/README.md [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+
+基于PARL库实现A2C算法，应用于[RLSchool][rlschool]库中的电梯调度模拟环境[LiftSim][liftsim]。
+
+## 依赖库
+
++ [paddlepaddle>=1.6.1](https://github.com/PaddlePaddle/Paddle)
++ [parl](https://github.com/PaddlePaddle/PARL)
++ [rlschool>=0.1.1][rlschool]
+
+Windows版本仅支持Python3.5及以上版本。
```
Comment (TomorrowIsAnOtherDay) [inline examples/LiftSim_baseline/A2C/README.md]: Our simulator is not supporting windows.

File: examples/LiftSim_baseline/A2C/README.md [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+```
+
+## 评估
+可以通过下面命令来评估保存的模型
+```bash
+python predict.py --model_path saved_models/[FILENAME]
+```
+
+## 收敛指标
+在GPU上训练30h左右，评估指标能达到-120分左右（LiftSim环境运行1天reward）
```
Comment (TomorrowIsAnOtherDay) [inline examples/LiftSim_baseline/A2C/README.md]: Remove GPU here. A majority of participants have limited computation resource.

File: examples/LiftSim_baseline/A2C/predict.py [hunk 3]
```diff
@@ -0,0 +1,61 @@
+#   Copyright (c) 2018 PaddlePaddle Authors. All Rights Reserved.
```
Comment (TomorrowIsAnOtherDay) [inline examples/LiftSim_baseline/A2C/predict.py]: rename this file to evaluate.py

File: examples/LiftSim_baseline/A2C/predict.py [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+
+    logger.info('model_path: {}, 24h reward: {}'.format(
+        model_path, reward_24h))
+
+
+if __name__ == '__main__':
+    import argparse
+    parser = argparse.ArgumentParser()
+    parser.add_argument(
+        '--model_path', type=str, help='path of the model to restore')
```
Comment (TomorrowIsAnOtherDay) [inline examples/LiftSim_baseline/A2C/predict.py]: to restore -> to evaluate

File: examples/LiftSim_baseline/DQN/README.md [hunk 5]
```diff
@@ -6,9 +6,9 @@
...
 ## 依赖库
 
-- paddlepaddle >= 1.5.1
-- parl >= 1.1.2
-- rlschool >= 0.0.1
+- paddlepaddle==1.5.1
+- parl==1.1.2
+- rlschool>=0.0.1
 
 Windows版本仅支持Python3.5及以上版本。
```
Comment (TomorrowIsAnOtherDay) [inline examples/LiftSim_baseline/DQN/README.md]: We are not supporting windows.

File: examples/LiftSim_baseline/A2C/actor.py [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+
+            for env_id in range(self.env_num * self.ele_num):
+                env_sample_data[env_id]['obs'].append(self.obs_batch[env_id])
+                env_sample_data[env_id]['actions'].append(
+                    actions_batch[env_id])
+                env_sample_data[env_id]['rewards'].append(reward_batch[env_id])
+                env_sample_data[env_id]['dones'].append(done_batch[env_id])
+                env_sample_data[env_id]['values'].append(values_batch[env_id])
+
+                # Calculate advantages when the episode is done or reach max sample steps.
```
Comment (TomorrowIsAnOtherDay) [inline examples/LiftSim_baseline/A2C/actor.py]: Calculate advantages when the episode is done or `reaches` max sample steps.

File: examples/LiftSim_baseline/DQN/README.md [hunk 7]
```diff
@@ -6,9 +6,9 @@
 
 ## 依赖库
 
-- paddlepaddle >= 1.5.1
-- parl >= 1.1.2
-- rlschool >= 0.0.1
+- paddlepaddle==1.5.1
+- parl==1.1.2
+- rlschool>=0.0.1
```
Comment (TomorrowIsAnOtherDay) [inline examples/LiftSim_baseline/DQN/README.md]: Please add a hyperlink for the rlschool repository here.

File: examples/LiftSim_baseline/A2C/utils.py [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+    active_pos = min(n_dim - 1, active_pos)
+    active_pos = max(0, active_pos)
+    ret_array = [0 for i in range(n_dim)]
+    ret_array[active_pos] = 1.0
+    return ret_array
+
+
+def linear_discretize(value, n_dim, min_val, max_val):
+    '''
+    discretize a value into a vector of n_dim dimensional representation
```
Comment (TomorrowIsAnOtherDay) [inline examples/LiftSim_baseline/A2C/utils.py:36]: represent a value with a one-hot vector.

File: examples/LiftSim_baseline/A2C/README.md (hunk 1 above)
Comment (zenghsh3) [reply examples/LiftSim_baseline/A2C/README.md]: removed

File: examples/LiftSim_baseline/A2C/README.md (hunk 2 above)
Comment (zenghsh3) [reply examples/LiftSim_baseline/A2C/README.md]: removed

File: examples/LiftSim_baseline/A2C/predict.py (hunk 3 above)
Comment (zenghsh3) [reply examples/LiftSim_baseline/A2C/predict.py]: done

File: examples/LiftSim_baseline/A2C/predict.py (hunk 4 above)
Comment (zenghsh3) [reply examples/LiftSim_baseline/A2C/predict.py]: done

File: examples/LiftSim_baseline/DQN/README.md (hunk 5 above)
Comment (zenghsh3) [reply examples/LiftSim_baseline/DQN/README.md]: removed

File: examples/LiftSim_baseline/A2C/actor.py (hunk 6 above)
Comment (zenghsh3) [reply examples/LiftSim_baseline/A2C/actor.py]: done

File: examples/LiftSim_baseline/DQN/README.md (hunk 7 above)
Comment (zenghsh3) [reply examples/LiftSim_baseline/DQN/README.md]: done

File: examples/LiftSim_baseline/A2C/utils.py (hunk 8 above)
Comment (zenghsh3) [reply examples/LiftSim_baseline/A2C/utils.py:36]: It is not one-hot vector.

File: examples/LiftSim_baseline/A2C/train.py [hunk 9]
```diff
@@ -0,0 +1,226 @@
...
+            model, vf_loss_coeff=config['vf_loss_coeff'])
+        self.agent = LiftAgent(algorithm, config)
+
+        if machine_info.is_gpu_available():
+            assert get_gpu_count() == 1, 'Only support training in single GPU,\
+                    Please set environment variable: `export CUDA_VISIBLE_DEVICES=[GPU_ID_TO_USE]` .'
+
+        else:
+            cpu_num = os.environ.get('CPU_NUM')
+            assert cpu_num is not None and cpu_num == '1', 'Only support training in single CPU,\
```
Comment (TomorrowIsAnOtherDay) [inline examples/LiftSim_baseline/A2C/train.py]: I'm not sure if the assertion on CPU_NUM is still necessary now, as we have set it in the parl.compile function.  Please have a check on it and remove it if possible. https://github.com/PaddlePaddle/PARL/blob/c848bda2fcef55438b774e77ffba47d71df14418/parl/core/fluid/plutils/compiler.py#L36

File: examples/LiftSim_baseline/A2C/train.py [hunk 10]
```diff
@@ -0,0 +1,226 @@
...
+
+            self.remote_count += 1
+            logger.info('Remote actor count: {}'.format(self.remote_count))
+
+            remote_thread = threading.Thread(
+                target=self.run_remote_sample, args=(params_queue, ))
+            remote_thread.setDaemon(True)
+            remote_thread.start()
+
+        logger.info('All remote actors are ready, begin to learn.')
```
Comment (TomorrowIsAnOtherDay) [inline examples/LiftSim_baseline/A2C/train.py]: When resource is scarce for actors, it is inappropriate to print this log.
I suggest deleting this log.

File: examples/LiftSim_baseline/A2C/README.md [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+```bash
+xparl start --port 8010 --cpu_num 5
+```
+
+> 注意，如果你已经启动了一个集群，则不需要重复运行上面命令。关于PARL集群更多信息，可以参考[文档](https://parl.readthedocs.io/en/latest/parallel_training/setup.html)。
+
+然后我们就可以通过运行下面命令进行分布式训练：
+
+```bash
+python train.py
```
Comment (TomorrowIsAnOtherDay) [inline examples/LiftSim_baseline/A2C/README.md:27]: Can you add the instruction about how to check the training performance with tensorboard here?

File: examples/LiftSim_baseline/A2C/train.py (hunk 10 above)
Comment (zenghsh3) [reply examples/LiftSim_baseline/A2C/train.py]: deleted

File: examples/LiftSim_baseline/A2C/train.py (hunk 9 above)
Comment (zenghsh3) [reply examples/LiftSim_baseline/A2C/train.py]: It is unnecessary now, removed.

File: examples/LiftSim_baseline/A2C/README.md (hunk 11 above)
Comment (zenghsh3) [reply examples/LiftSim_baseline/A2C/README.md:27]: done

**Review by TomorrowIsAnOtherDay — APPROVED** [review]
Looks good to me.

**Pull request merged**